src/avl.py

```diff
@@ -0,0 +1,468 @@
+import json
+import logging
+
+from hypothesis import given
+import hypothesis.strategies as st
+
+
+
+class AVLTree():
+    """The interface to an AVL tree.
+
+    The tree itself is actually represented by AVLTreeNode instances, but because of
+    needing to be able to change which node is the root, we need a container class.
+
+    Args:
+      value: an initial value to insert into the tree.
+    """
+
+    def __init__(self, value=None):
+        self._root = AVLTreeNode(self)
+        if value is not None:
+            self._root.insert(value)
+
+        self.log = []
+
+    def insert(self, value):
+        """Inserts the vallue into the tree.
+
+        Args:
+          value: the value to insert into the tree
+        """
+
+        self.root.insert(value)
+
+    def write_log(self, file_name="tree_log.json"):
+        """Writes a log of operation description and tree pairs in JSON for visualization."""
+
+        with open(file_name, "w") as f:
+            f.write(json.dumps(self.log))
+
+    def __str__(self):
+        return str(self.root)
+
+    @property
+    def root(self):
+        return self._root
+
+    @root.setter
+    def root(self, value):
+        self._root = value
+        self._root.parent = None
+
+    def to_dict(self):
+        """Returns a dictionary representation of this tree."""
+
+        return self._root.to_dict()
+
+    def __iter__(self):
+        return iter(self._root)
+
+
+class AVLTreeNode():
+    def __init__(self, container):
+        self.value = None
```

Comment:
I'm unclear on that too. I think there are larger stylistic concerns right now. This is sort of halfway between two stable and consistent styles, I think, and with time it'll hopefully become clear which is right. As of right now, I think I don't have enough information to say which is the right way to go, so I'll go with it until I've got some more experience using/extending this.